---
title: "human B cell atlas and deriving isotype signatures"
output: html_notebook
---
```{r setup}
library(Seurat)
library(sciCSR)
```

# Stewart et al (peripheral blood B cells)

```{r}
pure <- '/Volumes/Seagate4TB/GLT_datasets/scPure/Stewart_HumanPeripheral_Bcells_atlas_IGHC.rds'
pure <- readRDS(pure)
pure <- subset(pure, subset = (orig.ident == "HB6"))


#-------------------------
# Stewart et al Dotplot of Sterile/productive transcript expression
library(ggplot2)
genes <- rownames(pure@assays$IGHC)
genes <- genes[grepl("-[PS]$", genes)]

ighc_plot <- DotPlot(pure, assay = "IGHC", group.by = "library_id",
        features = genes) + coord_flip() + RotatedAxis()
ighc_plot <- ighc_plot$data
ighc_plot$features.plot <- factor(ighc_plot$features.plot,
                                  levels = genes)
ighc_plot$id <- factor(ighc_plot$id, levels = rev(c("Transitional",
                                                "Naive",
                                                "IgM_Memory",
                                                "Classical_Memory",
                                                "DN")))
svg("Documents/MACSMAF/sciCSR/Fig_scPure_IgH_dotplot.svg",
    width = 6.5, height = 2.5)
ggplot(ighc_plot[!grepl("IGHD", ighc_plot$features.plot), ],
       aes(y = id, x = features.plot, size = pct.exp,
           color = avg.exp.scaled)) +
  geom_point() + cowplot::theme_cowplot() +
  scale_color_gradient(low = "white", high = "purple", name = "Average\nExpression") +
  scale_size_continuous(name = "Percent\nExpressed") +
  scale_x_discrete(position = "top") +
  theme(axis.text.x = element_text(angle = 45, hjust = 0)) +
  ylab("") + xlab("")
dev.off()

```

## CSR potential

plot UMAP, CSR potential scores as feature plot, along side sterile IgG1/A1 just to show we can visualise these at single-cell level.

```{r}
pure <- FindNeighbors(pure)
pure <- getCSRpotential(pure, reference_based = "human")

df <- FetchData(pure, c("UMAP_1", "UMAP_2", "csr_pot"))
Idents(test) <- test$library_id

library(ggplot2)
pdf("Stewart_csr_pot.pdf", width = 17, height = 3.6)
cowplot::plot_grid(
  DimPlot(test, label = TRUE) + ggtitle("Stewart et al. Peripheral blood"),
  ggplot(df, aes(x = UMAP_1, y = UMAP_2, color = csr_pot)) +
    geom_point(data = df[which(df$csr_pot == 1), ], pch = ".") +
    # ensure points of different CSR potential values are plotted on top
    geom_point(data = df[which(df$csr_pot < 1), ], pch = ".") +
    cowplot::theme_cowplot() +
    scale_color_viridis_c(option = "C", name = "CSR\npotential"),
  FeaturePlot(test, features = c("IGHG1-S")),
  FeaturePlot(test, features = c("IGHA1-S")),
  nrow = 1, axis = "tblr", align = "hv"
)
dev.off()

```

# King et al tonsil dataset
```{r}
king <- '/Volumes//Seagate4TB/GLT_datasets/King_Tonsil/SEURAT_OBJECTS/King_HumanTonsil_Bcells_IGHC.rds'
king <- readRDS(king)
king <- subset(king, idents = c("Naive", "MBC", "Activated", "preGC", "LZ GC", "DZ GC", "prePB",
                                "Plasmablast"))
Idents(king) <- factor(Idents(king), levels = c("Naive", "MBC", "Activated", "preGC", "LZ GC", "DZ GC", "prePB",
                                                "Plasmablast"))
king$CellType <- factor(king$CellType, levels = c("Naive", "MBC", "Activated", "preGC", "LZ GC", "DZ GC", "prePB",
                                                  "Plasmablast"))
```

CSR potential

```{r}
king <- Seurat::FindNeighbors(king, assay = "integrated")
king <- RunUMAP(king, dims = 1:9) # at least 1.5% variance explaind in each PC
DefaultAssay(king) <- "integrated"
king <- sciCSR::getCSRpotential(king, reference_based = "human")
df <- FetchData(king, c("UMAP_1", "UMAP_2", "csr_pot"))

library(ggplot2)
pdf("King_csr_pot.pdf", width = 17, height = 3.6)
cowplot::plot_grid(
  # UMAP
  DimPlot(test, label = TRUE) + ggtitle("King et al. Tonsils"),
  # CSR potential feature plot (i.e. csr_pot = colour of the dots)
  ggplot(df, aes(x = UMAP_1, y = UMAP_2, color = csr_pot)) + 
    geom_point(data = df[which(df$csr_pot == 1), ], pch = ".") + 
    # ensure points of different CSR potential values are plotted on top
    geom_point(data = df[which(df$csr_pot < 1), ], pch = ".") + 
    cowplot::theme_cowplot() + 
    scale_color_viridis_c(option = "C", name = "CSR\npotential"),
  # sterile IgG1 feature plot
  FeaturePlot(test, features = c("IGHG1-S")),
  # sterile IgA1 feature plot
  FeaturePlot(test, features = c("IGHA1-S")),
  nrow = 1, axis = "tblr", align = "hv"
)
dev.off()

```

# Sparsity of productive/sterile transcripts

Counting % of cells positive for each productive/sterile transcript.

```{r}
#-------------------------
# % cells with zero counts distribution
# plot productive/sterile counts on top
pure_nnz <- tabulate(slot(pure[["RNA"]], "counts")@i + 1) / length(Cells(pure))
pure_nnz <- data.frame(nnz = pure_nnz)
pure_ighc_nnz <- data.frame(gene = rownames(slot(pure[["IGHC"]], "counts")),
                            nnz = tabulate(slot(pure[["IGHC"]], "counts")@i + 1) / 
                              length(Cells(pure)))
pure_ighc_nnz$type <- sapply(pure_ighc_nnz$gene, function(x) unlist(strsplit(x, split = "-"))[2])
pure_ighc_nnz$type <- factor(pure_ighc_nnz$type, levels = c("S", "P", "C"),
                             labels = c("sterile", "productive", "uninformative"))
pure_ighc_nnz$gene <- gsub("IGH", "", gsub("-.*", "", pure_ighc_nnz$gene))
pure_ighc_nnz <- pure_ighc_nnz[which(pure_ighc_nnz$gene != "D"), ]

king_nnz <- tabulate(slot(king[["RNA"]], "counts")@i + 1) / length(Cells(king))
king_nnz <- data.frame(nnz = king_nnz)
king_ighc_nnz <- data.frame(gene = rownames(slot(king[["IGHC"]], "counts")),
                            nnz = tabulate(slot(king[["IGHC"]], "counts")@i + 1) / 
                              length(Cells(king)))
king_ighc_nnz$type <- sapply(king_ighc_nnz$gene, function(x) unlist(strsplit(x, split = "-"))[2])
king_ighc_nnz$type <- factor(king_ighc_nnz$type, levels = c("S", "P", "C"),
                             labels = c("sterile", "productive", "uninformative"))
king_ighc_nnz$gene <- gsub("IGH", "", gsub("-.*", "", king_ighc_nnz$gene))
king_ighc_nnz <- king_ighc_nnz[which(king_ighc_nnz$gene != "D"), ]

library(ggplot2)
g_pure_nnz <- ggplot(pure_nnz, aes(nnz)) + geom_histogram() + 
  scale_x_log10(labels = scales::percent, name = "% cells with positive counts") +
  geom_point(aes(y = 1200, x = nnz), data = pure_ighc_nnz[which(pure_ighc_nnz$type == "productive"), ], 
             pch = 4, color = "#d45500") +
  geom_text(aes(y = 1200, x = nnz, label = gene), data = pure_ighc_nnz[which(pure_ighc_nnz$type == "productive"), ],
            vjust = -1.5, color = "#d45500") + 
  geom_text(aes(y = 1200, x = 0.65, label = "productive"), vjust = -1.5, color = "#d45500") + 
  geom_point(aes(y = 1400, x = nnz), data = pure_ighc_nnz[which(pure_ighc_nnz$type == "sterile"), ], 
             pch = 4, color = "#e5b799") +
  geom_text(aes(y = 1400, x = nnz, label = gene), data = pure_ighc_nnz[which(pure_ighc_nnz$type == "sterile"), ],
            vjust = -1.5, color = "#e5b799") + 
  geom_text(aes(y = 1400, x = 0.65, label = "sterile"), vjust = -1.5, color = "#e5b799") + 
  scale_y_continuous(breaks = c(0, 500, 1000), limits = c(0, 1500), name = "Number of genes") +
  cowplot::theme_cowplot()
g_king_nnz <- ggplot(king_nnz, aes(nnz)) + geom_histogram() + 
  scale_x_log10(labels = scales::percent, name = "% cells with positive counts") +
  geom_point(aes(y = 1600, x = nnz), data = king_ighc_nnz[which(king_ighc_nnz$type == "productive"), ], 
             pch = 4, color = "#d45500") +
  geom_text(aes(y = 1600, x = nnz, label = gene), data = king_ighc_nnz[which(king_ighc_nnz$type == "productive"), ],
            vjust = -1.5, color = "#d45500") + 
  geom_text(aes(y = 1600, x = 0.65, label = "productive"), vjust = -1.5, color = "#d45500") + 
  geom_point(aes(y = 1800, x = nnz), data = king_ighc_nnz[which(king_ighc_nnz$type == "sterile"), ], 
             pch = 4, color = "#e5b799") +
  geom_text(aes(y = 1800, x = nnz, label = gene), data = king_ighc_nnz[which(king_ighc_nnz$type == "sterile"), ],
            vjust = -1.5, color = "#e5b799") + 
  geom_text(aes(y = 1800, x = 0.65, label = "sterile"), vjust = -1.5, color = "#e5b799") + 
  scale_y_continuous(breaks = c(0, 500, 1000, 1500), limits = c(0, 2000), name = "Number of genes") +
  cowplot::theme_cowplot()

pdf("Documents/MACSMAF/sciCSR/sparsity.pdf", width = 5.5, height = 8 )
cowplot::plot_grid(
  g_pure_nnz + ggtitle("Stewart et al\n(peripheral blood)"), 
  g_king_nnz + ggtitle("King et al\n(tonsil)"), ncol = 1, align = "v", axis = "lr"
)
dev.off()
```

```{r}
#---------------------------
# count of cells with sterile transcripts 0/==1/>1 isotype
sterile_counts <- list(
  slot(pure[["IGHC"]], "counts"),
  slot(king[["IGHC"]], "counts")
)
sterile_counts <- lapply(sterile_counts, function(x){
  Matrix::t(x[which(grepl("-S", rownames(x)) & !grepl("IGHD", rownames(x))), ])
})
sterile_stats <- lapply(sterile_counts, function(x){
  y <- x@i + 1
  y <- factor(y, levels = 1:nrow(x))
  c(table(y))
})

pure_sterile <- pure@meta.data[, c("library_id"), drop = FALSE]
king_sterile <- king@meta.data[, c("CellType"), drop = FALSE]
sterile_stats <- do.call("rbind", list(
  data.frame(cell_type = pure_sterile$library_id, sterile_n = sterile_stats[[1]],
             dataset = "Stewart et al."),
  data.frame(cell_type = king_sterile$CellType, sterile_n = sterile_stats[[2]],
             dataset = "King et al.")
))
library(plyr)
sterile_stats <- merge(
  ddply(sterile_stats, c("dataset", "cell_type"),
        summarise,
        n_zero = sum(sterile_n == 0),
        n_one = sum(sterile_n == 1),
        n_plus = sum(sterile_n >1)),
  ddply(sterile_stats, c("dataset", "cell_type"), nrow),
  by = c("dataset", "cell_type")
)
sterile_stats$prop_zero <- sterile_stats$n_zero / sterile_stats$V1
sterile_stats$prop_one <- sterile_stats$n_one / sterile_stats$V1
sterile_stats$prop_plus <- sterile_stats$n_plus / sterile_stats$V1
sterile_stats <- reshape2::melt(sterile_stats, id.vars = c("dataset", "cell_type"), 
                                measure.vars = c("prop_zero", "prop_one", "prop_plus"))
sterile_stats[which(sterile_stats$dataset == "Stewart et al." &
                      sterile_stats$cell_type == "Naive"), "cell_type"] <- " Naive"
sterile_stats$cell_type <- factor(sterile_stats$cell_type,
                                  levels = c("Naive", "MBC", "Activated", "preGC", "LZ GC", "DZ GC", "prePB",
                                             "Plasmablast", "Transitional", " Naive",
                                             "IgM_Memory", "Classical_Memory", "DN"))
pdf("Documents/MACSMAF/sciCSR/sparsity_sterile.pdf", width = 6.2, height = 4 )
ggplot(sterile_stats, aes(x = cell_type, y = value, fill = variable)) +
  geom_bar(position = "fill", stat = "identity") + cowplot::theme_cowplot() +
  facet_grid(cols = vars(dataset), scales = "free_x", space = "free_x") +
  theme(axis.text.x = element_text(angle = 45, hjust = 1)) +
  scale_y_continuous(labels = scales::percent, name = "% cells") + xlab("") +
  scale_fill_manual(values = c("prop_zero" = "grey80", "prop_one" = "grey50",
                               "prop_plus" = "grey20"),
                    labels = c("no sterile\ntranscripts", "1 isotype", ">1 isotype"), 
                    name = "Sterile transcripts\nobserved in cell")
dev.off()
```

```{r}
#---------------------------
# total UMI count per cell vs total sterile transcript counts vs scaled
steriles <- paste0("IGH", c("M", "G3", "G1", "A1", "G2", "G4", 'E', "A2"), "-S")
pure <- AddMetaData(pure, Matrix::colSums(slot(pure[["IGHC"]], "counts")[steriles, ]),
                    "total_ic")
pure <- ScaleData(pure, vars.to.regress = "nCount_RNA", assay = "IGHC")
pure <- AddMetaData(pure, Matrix::colSums(slot(pure[["IGHC"]], "scale.data")[steriles, ]),
                    "scaled_total_ic")
pdf("Documents/MACSMAF/sciCSR/total_sterile_regression.pdf", width = 7.7, height = 4.1 )
cowplot::plot_grid(
  ggplot(pure@meta.data, aes(x = nCount_RNA, y = total_ic)) + 
    geom_point() + geom_smooth(method = "loess") + cowplot::theme_cowplot() +
    xlab("Total UMI count per cell") + 
    ylab("Total expression of sterile transcripts per cell") +
    ggtitle("No regression"),
  ggplot(pure@meta.data, aes(x = nCount_RNA, y = scaled_total_ic)) + 
    geom_point() + geom_smooth(method = "loess") + cowplot::theme_cowplot() +
    xlab("Total UMI count per cell") + 
    ylab("") + ggtitle("Regressed out Total UMI count"),
  nrow = 1, align = "h", axis = "tb"
)
dev.off()
```
# Combining the Stewart et al and King et al data

```{r}
# read Stewart et al & King et al atlas
pure <- 'Stewart_HumanPeripheral_Bcells_IGHC.rds'
pure <- readRDS(pure)
pure <- subset(pure, subset = (orig.ident == "HB6")) # the original donor
DefaultAssay(pure) <- 'integrated'

king <- 'King_HumanTonsil_Bcells_IGHC.rds'
king <- readRDS(king)

pure <- NormalizeData(pure)
pure <- FindVariableFeatures(pure, selection.method = "vst", nfeatures = 2000)
king <- NormalizeData(king)
king <- FindVariableFeatures(king, selection.method = "vst", nfeatures = 2000)
```

```{r}
# integrate two atlas into one for fitting trajectory using slingshot
features <- SelectIntegrationFeatures(object.list = list(pure, king))
anchors <- FindIntegrationAnchors(object.list = list(pure, king), anchor.features = features)
combined <- IntegrateData(anchorset = anchors)
```

```{r}
# Run the standard workflow for visualization and clustering
combined <- ScaleData(combined, verbose = FALSE)
combined <- RunPCA(combined, npcs = 30, verbose = FALSE)
combined <- RunUMAP(combined, reduction = "pca", dims = 1:30)
combined <- FindNeighbors(combined, reduction = "pca", dims = 1:30)
combined <- FindClusters(combined, resolution = 0.5)

DimPlot(combined, group.by = "orig.ident")
DimPlot(combined, group.by = "transfer_label", label = TRUE)
DimPlot(combined, group.by = "CellType", label = TRUE)
DimPlot(combined, label = TRUE)
```

```{r}
saveRDS(combined, '/media/josefng/Seagate4TB/GLT_datasets/human_Bcells_atlas_IGHC.rds')
```

# NMF to derive CSR potential

One easy way to derive a score would be to decompose the productive/sterile count matrix for "isotype signatures" - one would expect a prominent signature being a naive like signature (i.e. IgM-dominated) - and then extract the weight (coefficient) associated with this signature for each cell. And then invert this such that naive cells have a small number. The signature matrix can be stored and decompose future samples (i.e. new datasets) using nonnegative least square (NNLS) regression.

The matrix decomposition was performed using nonnegative matrix factorization (NMF). Here a simple NMF decomposition into 2 signatures (i.e. rank = 2) were performed - the idea is oen should expect a naive signature (heavy weights for IgM productive/sterile) and a memory signature. The safest seems to be to use the naive signature (as memory can be biased towards any isotype depending on the datasets, both reference data (i.e. our atlases)) or new datasets from the user).

```{r}
combined <- readRDS('human_Bcells_atlas_IGHC.rds')
```

```{r}
stewart <- subset(combined, subset = (orig.ident == "HB6"))
king <- subset(combined, subset = (Lineage == "B Cells"))
library(NMF)
normcounts <- list("stewart" = as.matrix(stewart@assays$IGHC@data),
                   "king" = as.matrix(king@assays$IGHC@data))
normcounts[["stewart"]] <- normcounts[["stewart"]][!grepl("^IGHD", rownames(normcounts[["stewart"]])), ]
normcounts[["king"]] <- normcounts[["king"]][!grepl("^IGHD", rownames(normcounts[["king"]])), ]
normcounts[["stewart"]] <- normcounts[["stewart"]][!grepl("-C$", rownames(normcounts[["stewart"]])), ]
normcounts[["king"]] <- normcounts[["king"]][!grepl("-C$", rownames(normcounts[["king"]])), ]
normcounts[["stewart"]] <- normcounts[["stewart"]][, colSums(normcounts[["stewart"]]) > 0]
normcounts[["king"]] <- normcounts[["king"]][, colSums(normcounts[["king"]]) > 0]
saveRDS(normcounts, "human_Bcells_atlas_IGHC_NormCounts.rds")
```

```{r}
normcounts <- readRDS("human_Bcells_atlas_IGHC_NormCounts.rds")
```

```{r}
# test NMF of different ranks. set upper bound = 8 as there are 8 isotypes except IgD
library(NMF)
test_nmf <- list()
test_nmf <-  nmf(normcounts, rank = 2:8, method = "snmf/r", 
                 nrun = 10, .options = "p4 v")
```


```{r}
# final rank=2 nmf
nmf <- nmf(normcounts, rank = 2, method = "snmf/r", nrun = 10, .options = "p2 v")
saveRDS(nmf, "human_Bcells_atls_IGHC_NMF_rank.rds")
```

```{r}
# this is the signature matrix
pheatmap::pheatmap( nmf@fit@W, cluster_rows = FALSE, cluster_cols = FALSE )

# the first column is the naive signature

# fit the coefficients of these two signature for each cell in the data
# using nnls

naive_signature <- sapply(Cells(combined),
                          function(n) nnls::nnls(nmf@fit@W, combined@assays$IGHC@data[rownames(nmf@fit@W), n])$x[1])
# scale into range [0, 1] and invert
naive_signature <- (naive_signature - min(naive_signature)) / abs(diff(range(naive_signature)))
naive_signature <- 1 - naive_signature
combined <- AddMetaData(combined, naive_signature, col.name = "csr_score")
```

### Check CSR potential against sterile/productive transcription levels

```{r}
pure <- AddMetaData(pure, FetchData(pure, "IGHM-P"), "IgM")
pure <- AddMetaData(pure, rowSums(FetchData(pure, paste0(c("IGHG1", "IGHG2", "IGHG3",
                                                           "IGHG4", "IGHA1", "IGHA2"), "-P"))), "IgG.A")
pure <- AddMetaData(pure, rowSums(FetchData(pure, paste0(c("IGHG1", "IGHG2", "IGHG3",
                                                           "IGHG4", "IGHA1", "IGHA2"), "-S"))), "IgG.A_sterile")

king <- subset(king, subset = (CellType %in% c("Naive", "GC", "Plasmablast")))
king <- FindNeighbors(king, assay = "integrated", return.neighbor = TRUE)
DefaultAssay(king) <- "integrated"
king <- getCSRpotential(king, reference_based = "human")
king <- AddMetaData(king, FetchData(king, "IGHM-P"), "IgM")
king <- AddMetaData(king, rowSums(FetchData(king, paste0(c("IGHG1", "IGHG2", "IGHG3",
                                                           "IGHG4", "IGHA1", "IGHA2"), "-P"))), "IgG.A")
king <- AddMetaData(king, rowSums(FetchData(king, paste0(c("IGHG1", "IGHG2", "IGHG3",
                                                           "IGHG4", "IGHA1", "IGHA2"), "-S"))), "IgG.A_sterile")

nmf <- readRDS("/media/josephn/Seagate4TB/GLT_datasets/scPure-HB6_KingTonsil_IGHC_NMF-combined.rds")
nmf_weights <- nmf@fit@H[1, grepl("^[ACTG]", colnames(nmf@fit@H))]
pure <- AddMetaData(pure, nmf_weights, col.name = "nmf_weights")
nmf_weights <- nmf@fit@H[1, !grepl("^[ACTG]", colnames(nmf@fit@H))]
king <- AddMetaData(king, nmf_weights, col.name = "nmf_weights")

#------------------------------
# plot CSR potential against sterile/productive isotype expression
test <- FetchData(pure, c("nmf_weights", "library_id", "IgM", "IgG.A", "IgG.A_sterile"))
test <- test[which(test$nmf_weights > 0), ]
test <- test[which(test$library_id != "DN"), ]
test <- list(test[, c("nmf_weights", "IgM")], test[, c("nmf_weights", "IgG.A")],
             test[, c("nmf_weights", "IgG.A_sterile")])
colnames(test[[1]]) <- c("nmf_weights", "expression")
colnames(test[[2]]) <- c("nmf_weights", "expression")
colnames(test[[3]]) <- c("nmf_weights", "expression")
test[[1]]$isotype <- "IgM"; test[[2]]$isotype <- "IgG/A"; test[[3]]$isotype <- "IgG/A_sterile"
test <- rbind(test[[1]], test[[2]], test[[3]])

test2 <- FetchData(pure, c("library_id", "nmf_weights"))
test2$library_id <- factor(test2$library_id,
                           levels = c("Transitional", "Naive", "IgM_Memory",
                                      "Classical_Memory", "DN"))
test2 <- test2[which(test2$library_id != "DN"), ]

library(ggplot2)
svg("nmf_weights_scPure.svg", width = 4.8, height = 4.6)
cowplot::plot_grid(
  ggplot(test2[which(test2$nmf_weights > 0), ],
         aes(x = nmf_weights, color = library_id)) +
    geom_density(adjust = 3, size = 1) +
    cowplot::theme_cowplot() + scale_x_reverse() +
    scale_color_manual(values = viridis::inferno(4)[1:4]) +
    geom_boxplot(aes(y = 0.8, fill = library_id), color = "grey", outlier.shape = NA,
                 position = position_dodge2(width = 0.3), width = 0.3) +
    scale_fill_manual(values = viridis::inferno(4)[1:4]) +
    scale_y_continuous(breaks = c(0, 0.2, 0.4, 0.6), name = "density"),
  ggplot(test, aes(x = nmf_weights, y = expression, color = isotype)) +
    geom_smooth(method = "loess") + cowplot::theme_cowplot() + scale_x_reverse(),
  align = "v", axis = "lr", ncol = 1
)
dev.off()

#----------------------------------
# plot the same for king et al GC
test <- FetchData(king, c("nmf_weights", "CellType", "IgM", "IgG.A", "IgG.A_sterile"))
test <- test[which(test$nmf_weights > 0), ]
test <- list(test[, c("nmf_weights", "IgM")], test[, c("nmf_weights", "IgG.A")],
             test[, c("nmf_weights", "IgG.A_sterile")])
colnames(test[[1]]) <- c("nmf_weights", "expression")
colnames(test[[2]]) <- c("nmf_weights", "expression")
colnames(test[[3]]) <- c("nmf_weights", "expression")
test[[1]]$isotype <- "IgM"; test[[2]]$isotype <- "IgG/A"; test[[3]]$isotype <- "IgG/A_sterile"
test <- rbind(test[[1]], test[[2]], test[[3]])

test2 <- FetchData(king, c("CellType", "nmf_weights"))

library(ggplot2)
svg("nmf_weights_king.svg", width = 4.8, height = 4.6)
cowplot::plot_grid(
  ggplot(test2[which(test2$nmf_weights > 0), ],
         aes(x = nmf_weights, color = CellType)) +
    geom_density(adjust = 3, size = 1) +
    cowplot::theme_cowplot() + scale_x_reverse() +
    scale_color_manual(values = viridis::inferno(3)) +
    geom_boxplot(aes(y = 0.8, fill = CellType), color = "grey", outlier.shape = NA,
                 position = position_dodge2(width = 0.3), width = 0.3) +
    scale_fill_manual(values = viridis::inferno(3)) +
    scale_y_continuous(breaks = c(0, 0.2, 0.4, 0.6), name = "density"),
  ggplot(test, aes(x = nmf_weights, y = expression, color = isotype)) +
    geom_smooth(method = "loess") + cowplot::theme_cowplot() + scale_x_reverse(),
  align = "v", axis = "lr", ncol = 1
)
dev.off()

```

## Mouse

```{r}
mouse <- readRDS("mouse_Bcells_atlas_IGHC.rds")
normcounts <- as.matrix(mouse@assays$IGHC@data)
normcounts <- normcounts[!grepl("^Ighd", rownames(normcounts)), ]
normcounts <- normcounts[!grepl("-C$", rownames(normcounts)), ]
normcounts <- normcounts[, colSums(normcounts) > 0]
saveRDS(normcounts, "mouse_Bcells_atlas_IGHC_NormCounts.rds")

# run nmf
library(NMF)
set.seed(1234)
test <- normcounts[,sample(1:ncol(normcounts), 40000)]
nmf <- nmf(test, rank = 2, method = "snmf/r", nrun = 10, .options = "p2 v")
saveRDS(nmf, "mouse_Bcells_atlas_IGHC_NMF_rank.rds")
```
