# sciCSR-analysis
analysis scripts in the Ng et al. sciCSR paper.

| filename | Description | Relevant figure(s) |
|----------|-------------|--------------------|
| folder `simulation` | Bash code used to process simulated IgH reads | Figure 2b,c |
| `plotBedGraphs.Rmd` | plotting distribution of reads & motifs across IgH locus | Figure 2a, Supplementary Figures S1, S2 |
| `simulate_IGHC_reads.Rmd` | Code to simulate RNA-seq reads from IgH transcripts and analysis | Figure 2b,c |
| `human_atlas.Rmd` | Code to assemble reference atlas for human B cells, visualisation of productive/sterile transcription, derivation of isotype signatures | Figure 2d, 4a, 4b, Supplementary Figure S3, S5, S6, S7 |
| `scIFNg_sciCSR.Rmd` | Code to analyse the interferon gamma culture scRNA-seq/scBCR-seq data | Figure 3 |
| `Kim_Covid_GLT.Rmd` | Code to analyse the SARS-CoV-2 vaccination time-course ([Kim et al. Nature 2022](https://www.nature.com/articles/s41586-022-04527-1)) data | Figure 5, Supplementary Figure S4, S8 |
| `Gomez_AID.Rmd` | Code to analyse the Aicda mouse knockout ([GÃ³mez-Escolar et al. EMBO Rep 2022](https://www.embopress.org/doi/full/10.15252/embr.202255000)) data | Figure 6a,b |
| `Hong_IL23_GLT.Rmd` | Code to analyse the Il23 mouse knockout ([Hong et al. J Immunol 2020](https://doi.org/10.4049/jimmunol.2000280)) data | Figure 6a,c-f, Supplementary Figure S9 |

